cmd/shell.go

```diff
@@ -0,0 +1,109 @@
+package cmd
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+// sigCh is called to shutdown the server.
+func runShell(mountpath string, sigCh chan os.Signal) exitCode {
+	cleanup := func(msgs ...interface{}) exitCode {
+		log.Print(msgs...)
+		sigCh <- syscall.SIGINT
+		return exitCode{1}
+	}
+
+	// Run the default system shell.
+	sh := os.Getenv("SHELL")
+	if sh == "" {
+		sh = "/bin/sh"
+	}
+
+	// Add wash to PATH, set PS1. These are useful fallbacks in case PATH and prompt aren't setup.
+	washPath, err := os.Executable()
+	if err != nil {
+		return cleanup(err)
+	}
+	washDir := filepath.Dir(washPath)
+	pathEnv := os.Getenv("PATH")
+	if err := os.Setenv("PATH", pathEnv+string(os.PathListSeparator)+washDir); err != nil {
+		return cleanup(err)
+	}
+
+	if err := os.Setenv("PS1", "wash$ "); err != nil {
+		return cleanup(err)
+	}
+
+	comm := exec.Command(sh)
+	comm.Stdin = os.Stdin
+	comm.Stdout = os.Stdout
+	comm.Stderr = os.Stderr
+	comm.Dir = mountpath
+	runErr := comm.Run()
+
+	// Shutdown the server
+	sigCh <- syscall.SIGINT
+
+	if runErr != nil {
+		if exitErr, ok := runErr.(*exec.ExitError); ok {
+			return exitCode{exitErr.ExitCode()}
+		}
+		log.Print(runErr)
+		return exitCode{1}
+	}
+	return exitCode{0}
+}
+
+// Start the wash server, then present the default system shell.
+// On exit, stop the server and return any errors.
+func shellMain(cmd *cobra.Command, args []string) exitCode {
+	cachedir, err := os.UserCacheDir()
+	if err != nil {
+		log.Print("Unable to get user cache dir:", err)
+		return exitCode{1}
+	}
+	cachedir = filepath.Join(cachedir, "wash")
+
+	// ensure cache directory exists
+	if err = os.MkdirAll(cachedir, 0750); err != nil {
+		log.Print("Unable to create cache dir:", cachedir)
+		return exitCode{1}
+	}
+
+	mountpath, err := ioutil.TempDir(cachedir, "mnt")
+	if err != nil {
+		log.Print("Unable to create temporary mountpoint:", err)
+		return exitCode{1}
+	}
+	defer os.RemoveAll(mountpath)
+
+	readyCh := make(chan struct{}, 1)
+	sigCh := make(chan os.Signal, 1)
+	doneCh := make(chan struct{}, 1)
+	go func() {
+		// Configure logrus to omit simple text
+		logrus.SetLevel(logrus.WarnLevel)
+		logrus.SetFormatter(&logrus.TextFormatter{DisableTimestamp: true})
+
+		// TODO: handle multiple active shells using a single server
```

Comment:
Not sure what you mean by this TODO.

Reply:
Also, how can I access my server logs?

Reply:
The server logs don't exist. They're written to stdout, or are part of activity journals.

Reply:
So if I want to debug something, I'd need to use the old workflow of starting my server in one shell, then doing stuff in another shell?

Reply:
No. Logging still goes to the console. I haven't added a flag to mess with verbosity yet.

Reply:
That could cause interleaved log/shell output, which might be confusing. Maybe include the logfile as an option too?